### PR TITLE
cli refactor: move model out of direct commands

### PR DESCRIPTION
**Pull request by dkrieger-pvtl**
We want to have a "view" that is separated from the "model" presented
by the underlying commands to enable cleaner testing and refactoring.

### CODE REVIEW

**Comment by dkrieger-pvtl** [comment]
Well, my overall motivation frankly was to do something straightforward to learn some golang while improving the code.  IMHO, the cli command structure itself is fairly complicated, so I wanted to separate out anything that wasn't purely just that.  Yes, this entire layer IS the cli, but to me it's easier to understand with this fix.  Certainly that's a matter of opinion; the main engineering argument is that I'd rather be dealing with logical functions and not string comparisons throughout as it is less fragile.   But this is a single component.

**Comment by jchampio** [comment]
> the main engineering argument is that I'd rather be dealing with logical functions and not string comparisons throughout as it is less fragile.

No argument from me there. But we can make that change without a new layer of code.

File: cli/commanders/reporter.go [hunk 1]
```diff
@@ -91,46 +98,59 @@ func (s StepStatuses) Swap(i, j int) {
...
 	statuses := conversionStatus.GetConversionStatuses()
 	sort.Sort(PrimaryStatuses(statuses))
-	formatStr := "%s - DBID %d - CONTENT ID %d - PRIMARY - %s"
-
+	formatStr := "%s - DBID %d - CONTENT ID %d - PRIMARY - %s\n"
+	stringBuf := strings.Builder{}
 	for _, status := range statuses {
 		reportString := fmt.Sprintf(formatStr, status.Status, status.Dbid, status.Content, status.Hostname)
-		fmt.Println(reportString)
+		stringBuf.WriteString(reportString)
```
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply cli/commanders/reporter.go]: Yes, it passes ```make check```.  I added a ```\n``` to the format string.

File: cli/commanders/check_config.go [hunk 2]
```diff
@@ -2,23 +2,27 @@ package commanders
...
+}
+
 func NewConfigChecker(client pb.CliToHubClient) ConfigChecker {
 	return ConfigChecker{
 		client: client,
 	}
 }
 
-func (req ConfigChecker) Execute() error {
+func (req ConfigChecker) Execute()  error {
```
[the remark this answers is not in the document]
Comment (dkrieger-pvtl) [reply cli/commanders/check_config.go]: Thanks; done.  I've repushed with 'gofmt' having run.....

**Comment by dkrieger-pvtl** [comment]
I think my intent is justifiable: to allow the "view" and "model" to be separated.  It's true, however, that the cli is the "view" and the hub is the "model" from this standpoint.  My main issue overall is that the cli's view is tightly coupled to the hub's model.  It's technically not because we can implement the gRPC on the hub in many ways, but in practice that's how we implement it.

I'd prefer to see a more model-driven view. That is, a way for the the cli to abstractly request the steps to do from the hub, and not need to know in detail how the hub does it.  Our upgrade steps need to be tightly choreographed but the cli code doesn't need to know that.  For instance, something like the calls ```hub_getUpgradeSteps(), hub_getCurrentStep(), hub_getStatus(), hub_perform_next_step()``` etc.   Again, it's not clear this is needed just for this use case of gpupgrade, but I think it would make testing easier.

As for the PR, I could go either way with the extra layer.  But I do want to modify the stdout printing in the commands that go straight to the Cobra layer to return Golang data structures(maps, etc) to make testing less fragile.  At least for the non-bats tests.  The bats tests necessarily test the output directly, though I might add a "123ABC_IGNORE" filter in the bats test to allow dev print and the tests to still pass.

**Comment by doty-pivotal** [comment]
I took a look, but I don't have anything helpful to add at this time.

**Comment by dkrieger-pvtl** [comment]
Closed this PR, as we are doing a major refactor that no longer requires this approach.

**Pull request closed without merging**